cmake/Platform/Core/BoardPropertiesReader.cmake

```diff
@@ -0,0 +1,89 @@
+    if (${BOARD_NAME}.menu.CPUS)
+        if (BOARD_CPU)
+            LIST(FIND ${BOARD_NAME}.menu.CPUS ${BOARD_CPU} CPU_INDEX)
+            if (CPU_INDEX EQUAL -1)
+                message(FATAL_ERROR "Invalid BOARD_CPU (valid cpus: ${${BOARD_NAME}.menu.CPUS}).")
+            endif()
+        else()
+            message(FATAL_ERROR "Board has multiple CPU versions (${${BOARD_NAME}.menu.CPUS}). BOARD_CPU must be defined for target ${TARGET_NAME}.")
+        endif()
+        set(${OUTPUT_VAR} ${BOARD_NAME}.${BOARD_CPU} PARENT_SCOPE)
+    else()
+        set(${OUTPUT_VAR} ${BOARD_NAME} PARENT_SCOPE)
+    endif()
+endfunction()
+
+#=============================================================================#
+# _get_board_property
+# [PRIVATE/INTERNAL]
+#
+# _get_board_property(BOARD_ID PROPERTY_NAME OUTPUT_VAR)
+#
+#        BOARD_ID - return value from function "_get_board_id (BOARD_NAME, BOARD_CPU)". It contains BOARD_NAME and BOARD_CPU
+#        PROPERTY_NAME - property name for the board, eg.: bootloader.high_fuses
+#        OUT_VAR - variable holding value for the property
+#
+# Gets board property.
+# Reconstructs BOARD_NAME and BOARD_CPU from BOARD_ID and tries to find value at ${BOARD_NAME}.${PROPERTY_NAME},
+# if not found than try to find value at ${BOARD_NAME}.menu.cpu.${BOARD_CPU}.${PROPERTY_NAME}
+# if not found that show fatal error
+#=============================================================================#
+function(_GET_BOARD_PROPERTY BOARD_ID PROPERTY_NAME OUTPUT_VAR)
+    string(REPLACE "." ";" BOARD_INFO ${BOARD_ID})
+    list(GET BOARD_INFO 0 BOARD_NAME)
+    set(VALUE ${${BOARD_NAME}.${PROPERTY_NAME}})
+    if(NOT VALUE)
+        list(LENGTH BOARD_INFO INFO_PARAMS_COUNT)
+        if (${INFO_PARAMS_COUNT} EQUAL 2)
+            list(GET BOARD_INFO 1 BOARD_CPU)
+            VALIDATE_VARIABLES_NOT_EMPTY(VARS BOARD_CPU MSG "cannot find CPU info, must define BOARD_CPU.")
+            set(VALUE ${${BOARD_NAME}.menu.cpu.${BOARD_CPU}.${PROPERTY_NAME}})
+        endif()
+    endif()
+    if (NOT VALUE)
+        message(FATAL_ERROR "Board info not found: BoardName='${BOARD_NAME}' BoardCPU='${BOARD_CPU}' PropertyName='${PROPERTY_NAME}'")
+    endif()
+    set(${OUTPUT_VAR} ${VALUE} PARENT_SCOPE)
+endfunction()
+
+#=============================================================================#
+# _get_board_property_if_exists
+# [PRIVATE/INTERNAL]
+#
+# _get_board_property_if_exists(BOARD_ID PROPERTY_NAME OUTPUT_VAR)
+#
+#        BOARD_ID - return value from function "_get_board_id (BOARD_NAME, BOARD_CPU)". It contains BOARD_NAME and BOARD_CPU
+#        PROPERTY_NAME - property name for the board, eg.: bootloader.high_fuses
+#        OUT_VAR - variable holding value for the property
+#
+# Similar to _get_board_property, except it returns empty value if value was not found.
+#=============================================================================#
+function(_GET_BOARD_PROPERTY_IF_EXISTS BOARD_ID PROPERTY_NAME OUTPUT_VAR)

```

Comment:
Should be named `try_get_board_property` as this is the accepted convention for *soft* retrievals.
For example: In C# a `Dictionary<TKey,TValue>` is a data structure that accept Key-Value pairs. 
Attempting to get a value for a key that doesn't exist in the dictionary results in an `Exception`. If you need to avoid this behavior and simply check for the key's existence **before** retrieving the associated value, you'd call the `try_get_value` method instead - It would return a `boolean` indicating whether the operation has been successful or not and set the retrieved value if it did. 
It's very similar to what your'e attempting to accomplish here.

Reply:
Good point, you can rename it to *_try_get_board_property*, but i would like to encourage using underscore for internal functions.

Reply:
Yes of course, I did mean `_try_get_property`, forgot the underline 😆